### DISCUSSION TITLE
Running scripts on gpu or HPC

### DISCUSSION THREAD

**Discussion by kumaryash7** [pybamm-team/PyBaMM#3471] [Q&A]
Hi, I am trying to run my PyBaMM scripts on GPU or also HPC. Is there any modules to explain that? or maybe someone who had already tried doing so? I am unable to reduce the computation time.

Thank you in advance.

**Comment by agriyakhetarpal** [comment]
Hi @kumaryash7, the `23.9rc0` pre-release version of PyBaMM now includes support for a variant of the Jax solver that provides GPU execution (#3121). You would require installing `jax` and `jaxlib` with CUDA enabled on a Linux or a Windows Subsystem for Linux (WSL2) machine. If you are on a macOS system, you would need to install the `jax-metal` bridge via the official Apple developer instructions. On Windows, GPU support is possible through community-maintained unofficial wheels which require an external CUDA and CuDNN installation. There are a few issues (#3371, #3422, and #3443) which will soon be resolved; you may use the unreleased version with an installation of PyBaMM from source if you wish to access this feature. The Jax solver does not require the compilation of the build-time requirements (they may be skipped if you do not want to make use of the IDAKLU solver).

**Reply by kumaryash7** [reply]
Hi, Thank you the reply,
Can you share some working examples for the same for either MacOS or windows system?

**Reply by agriyakhetarpal** [reply]
Hi @kumaryash7, the access to the GPU functionality is opt-in and does not require any change to the code snippets you might be using with the Jax solver, since Jax selects the CPU or GPU/TPU backends based on whatever is available. The instructions on Windows for Jax with GPU support are available at https://github.com/cloudhan/jax-windows-builder, and the macOS-specific guide is available at https://developer.apple.com/metal/jax/. A minimal code snippet is here (from #3121): 
 
 ```python
import pybamm
import numpy as np

model = pybamm.lithium_ion.DFN()
model.convert_to_format = 'jax'
model.events = []  # remove events (not supported in jax)
geometry = model.default_geometry
param = model.default_parameter_values
param.update({"Current function [A]": "[input]"})
param.process_geometry(geometry)
param.process_model(model)

n = 10
k = 5
values = np.linspace(0.1, 0.5, 100)
var = pybamm.standard_spatial_vars
var_pts = {var.x_n: n, var.x_s: n, var.x_p: n, var.r_n: k, var.r_p: k}
mesh = pybamm.Mesh(geometry, model.default_submesh_types, var_pts)
disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
disc.process_model(model)
t_eval = np.linspace(0, 3600, 100)
solver = pybamm.JaxSolver(atol=1e-6, rtol=1e-6, method="BDF")
inputs = [{"Current function [A]": value} for value in values]
solution = solver.solve(model, t_eval, inputs=inputs)
```

**Reply by kumaryash7** [reply]
I am trying to run the code on a gpu node on the HPC but it is not detecting a GPU on the node.
I tried to run the same code on the HPC but still it is not detecting a GPU.

**Reply by agriyakhetarpal** [reply]
It should work without additional configuration on your end on the source installation of PyBaMM. Do you have CUDA installed from the Jax wheels?

**Reply by kumaryash7** [reply]
Running on GPU does not reduce my computational time. Is there anything else I should consider to reduce my computational time?

**Reply by agriyakhetarpal** [reply]
Hi @kumaryash7, this might be better answered if an MWE can be provided. I would also suggest opening a new discussion about this, since this discussion was more about how to run scripts on an available GPU, and what you are asking is about optimising them.

**Reply by Hanyu-B** [reply]
Hi @kumaryash7, I tried to use parallel computing to accelerate the simulation process on HPC, but my jobs are independent so that I can just use multiple CPU cores to run all the jobs simultaneously. I am also going to try to use GPU to see whether it may help accelerate each independent job.

**Reply by kumaryash7** [reply]
Hi @Hanyu-B, I was trying to parallelise my pybamm scripts on the cluster but could not. Do you have any success with it?

**Reply by Hanyu-B** [reply]
for CPU cluster, yes. For GPU clusters, I haven't tried because the current simulation time is acceptable.

**Reply by kumaryash7** [reply]
Hi, I am really interested in the work. Is there a way we could get in touch?

**Reply by Hanyu-B** [reply]
Hi, thanks for your interest and  my email is baihanyu27@gmail.com

**Reply by Hanyu-B** [reply]
from joblib import Parallel, delayed
def pybamm_simulation(soc_i,charging_protocol_i):
    """
    the code need to run and just save all the results 
    """
    return 

if __name__ == '__main__':
    soc_ini=[] # input 1
    charging_protocol=[] # input 2
    r=Parallel(n_jobs=8)(delayed(pybamm_simulation)(soc_i,charging_protocol_i) for soc_i,charging_protocol_i in zip(soc_ini,charging_protocol))

Here is a mini working example, and I just upload the code our HPC systems and request as many CPUs as I put for n_jobs.
